src/cliffordsynthesis/encoding/STQGatesEncoder.cpp

```diff
@@ -0,0 +1,236 @@
+  DEBUG() << "Asserting gate consistency";
+  for (std::size_t t = 0U; t < T / 2U; ++t) {
+    // asserting only a single gate is applied on each qubit.
+    for (std::size_t q = 0U; q < N; ++q) {
+      LogicVector singleQubitGateVariables{};
+      LogicVector twoQubitGateVariables{};
+      vars.collectSingleQubitGateVariables(t, q, singleQubitGateVariables);
+      // TODO: may need to return the vars back
+      collectTwoQubitGateVariables(t, q, true, twoQubitGateVariables);
+      collectTwoQubitGateVariables(t, q, false, twoQubitGateVariables);
+
+      IF_PLOG(plog::verbose) {
+        TRACE() << "Single Qubit Gate variables at time " << t << " and qubit "
+                << q;
+        for (const auto& var : singleQubitGateVariables) {
+          TRACE() << var.getName();
+        }
+
+        TRACE() << "Two Qubit Gate variables at time " << t << " and qubit "
+                << q;
+        for (const auto& var : twoQubitGateVariables) {
+          TRACE() << var.getName();
+        }
+      }
+
+      assertExactlyOne(singleQubitGateVariables);
+      assertExactlyOne(twoQubitGateVariables);
+    }
+  }
+}
+
+void encoding::STQGatesEncoder::assertGateConstraints() {
+  DEBUG() << "Asserting gate constraints";
+  for (std::size_t t = 0U; t < T; ++t) {
+    TRACE() << "Asserting gate constraints at time " << t;
+
+    const std::size_t pos = t < T / 2 ? t : t % (T / 2);
+    t % 2U == 0 ? assertSingleQubitGateConstraints(pos)
+                : assertTwoQubitGateConstraints(pos);
+
+    TRACE() << "Asserting r changes at time " << t;
+    lb->assertFormula(tvars->r[t + 1] == rChanges);
+  }
+}
+
+void encoding::STQGatesEncoder::assertSingleQubitGateConstraints(
+    const std::size_t pos) {
+  for (std::size_t q = 0U; q < N; ++q) {
+    assertZConstraints(pos, q);
+    assertXConstraints(pos, q);
+    assertRConstraints(pos, q);
+  }
+}
+
+void STQGatesEncoder::assertRConstraints(const std::size_t pos,
+                                         const std::size_t qubit) {
+  for (const auto gate : SINGLE_QUBIT_GATES_FOR_STQ_ENCODING) {
+    rChanges =
+        rChanges ^ LogicTerm::ite(vars.gS[pos][gateToIndex(gate)][qubit],
+                                  tvars->singleQubitRChange(pos, qubit, gate),
+                                  LogicTerm(0, static_cast<std::int16_t>(S)));
+  }
+}
+
+void encoding::STQGatesEncoder::assertTwoQubitGateConstraints(
+    const std::size_t pos) {
+  const auto& twoQubitGates = vars.gC[pos];
+  for (std::size_t ctrl = 0U; ctrl < N; ++ctrl) {
+    for (std::size_t trgt = 0U; trgt < N; ++trgt) {
+      if (ctrl == trgt) {
+        continue;
+      }
+      const auto changes = createTwoQubitGateConstraint(pos, ctrl, trgt);
+      lb->assertFormula(LogicTerm::implies(twoQubitGates[ctrl][trgt], changes));
+
+      DEBUG() << "Asserting CNOT on " << ctrl << " and " << trgt;
+    }
+  }
+}
+
+LogicTerm encoding::STQGatesEncoder::createTwoQubitGateConstraint(
```

Comment:
If I'm not missing something here, this method can be removed as it is just a duplicated from the base class.